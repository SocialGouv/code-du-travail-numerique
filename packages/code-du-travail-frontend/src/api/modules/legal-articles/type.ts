import {
  DocumentElasticWithSource,
  LaborCodeDoc,
} from "@socialgouv/cdtn-types";

export type DocumentElasticResult<T> = T & {
  _id: string;
};


export type ElasticLaborCodeArticle = DocumentElasticWithSource<
  Omit<LaborCodeDoc, "cid">
>;
