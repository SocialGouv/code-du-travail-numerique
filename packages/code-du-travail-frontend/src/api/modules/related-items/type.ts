import { SOURCES } from "@socialgouv/cdtn-utils";
import { DocumentElastic } from "@socialgouv/cdtn-types/build/elastic/common";

export const sources = [
  SOURCES.TOOLS,
  SOURCES.SHEET_SP,
  SOURCES.SHEET_MT,
  SOURCES.LETTERS,
  SOURCES.CONTRIBUTIONS,
  SOURCES.EXTERNALS,
] as const;

export type RelatedItem = Pick<DocumentElastic, "title"> & {
  source: (typeof sources)[number];
  url: string;
};
