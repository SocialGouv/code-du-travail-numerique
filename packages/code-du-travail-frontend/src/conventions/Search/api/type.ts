export type Agreement = {
  id: string;
  num: number;
  shortTitle: string;
  slug: string;
  title: string;
  highlight: {
    title: string;
    content: string;
    searchInfo: string;
  };
};
