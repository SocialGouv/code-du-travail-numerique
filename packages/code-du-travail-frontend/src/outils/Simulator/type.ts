export type Step<Name extends string> = {
  name: Name;
  label: string;
  Component: () => JSX.Element;
  options?: {
    isForm?: boolean;
    annotation?: JSX.Element;
    hasNoMarginBottom?: boolean;
  };
};

export type SimulatorState<StepName extends string> = {
  currentStepIndex: number;
};

export type SimulatorActions = {
  nextStep: () => void;
  previousStep: () => void;
};

export type SimulatorStore = SimulatorState<string> & SimulatorActions;
