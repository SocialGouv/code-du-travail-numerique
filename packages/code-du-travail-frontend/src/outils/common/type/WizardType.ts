import { FormApi } from "final-form";
import React from "react";

export interface Step {
  component: (props: WizardStepProps) => JSX.Element;
  label: string;
  name: string;
  skip?: (values: FormContent) => boolean;
}

export interface State {
  steps: Array<Step>;
  stepIndex: number;
}

export enum ActionName {
  reset = "reset",
  setStepIndex = "setStepIndex",
}

export type Action =
  | { type: ActionName.reset }
  | { type: ActionName.setStepIndex; payload: number };

export interface WizardStepProps {
  form: FormApi<FormContent>;
  dispatch: React.Dispatch<Action>;
}

interface ConventionCollective {
  id: string;
  num: number;
  shortTitle: string;
  slug: string;
  title: string;
}

export type FormContent = Record<string, string> & {
  ccn?: ConventionCollective;
  seniorityGreaterThanTwoYears: boolean;
  infos?: Record<string, string>;
};
