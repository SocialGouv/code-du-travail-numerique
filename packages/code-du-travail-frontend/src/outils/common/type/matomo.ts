export enum MatomoCommonEvent {
  TRACK_EVENT = "trackEvent",
  OUTIL = "outil",
}

export enum MatomoConventionCollectiveEvent {
  CC_TREATED = "cc_select_traitée",
  CC_UNTREATED = "cc_select_non_traitée",
}

export enum MatomoPreavisRetraiteEvent {
  MISE_RETRAITE = "mise",
  DEPART_RETRAITE = "depart",
  ANCIENNETE_PLUS_2_ANS = "anciennete_plus_2_ans",
  ANCIENNETE_MOINS_2_ANS = "anciennete_moins_2_ans",
}
