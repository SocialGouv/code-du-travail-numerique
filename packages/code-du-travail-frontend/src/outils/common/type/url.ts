export enum TrackUrlMatomo {
  PREAVIS_RETRAITE = "/outils/preavis-retraite",
}
