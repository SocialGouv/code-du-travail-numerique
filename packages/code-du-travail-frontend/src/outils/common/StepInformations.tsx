import {
  heuresRechercheEmploiData,
  preavisDemissionData,
  preavisLicenciementData,
} from "@cdt/data";
import { FormApi } from "final-form";
import React, { useMemo } from "react";

import Html from "../../common/Html";
import { MatomoActionEvent, trackQuestion } from "../../lib";
import { SelectQuestion } from "./SelectQuestion";
import {
  filterSituations,
  getFormProps,
  getNextQuestionKey,
  getOptions,
  getPastQuestions,
  getSituationsFor,
} from "./situations.utils";
import { SectionTitle } from "./stepStyles";
import { FormContent } from "./type/WizardType";

type OpenArray = Array<{
  key: string;
  status: boolean;
}>;

export type StepInformationsProps = {
  form: FormApi<FormContent>;
  actionEvent?: MatomoActionEvent;
};

export function StepInformations({
  form,
  actionEvent,
}: StepInformationsProps): JSX.Element {
  const { values } = form.getState();
  const { ccn, typeRupture = null, criteria = {} } = values;
  const idcc = ccn?.selected?.num ?? 0;

  const subLabel = (key) =>
    key === "ancienneté"
      ? "Choisissez parmi les catégories d'ancienneté telles que définies par la convention collective"
      : undefined;

  const questions = useMemo(() => {
    switch (actionEvent) {
      case MatomoActionEvent.HEURE_RECHERCHE_EMPLOI:
        return heuresRechercheEmploiData.questions;
      case MatomoActionEvent.PREAVIS_DEMISSION:
        return preavisDemissionData.questions;
      case MatomoActionEvent.PREAVIS_LICENCIEMENT:
        return preavisLicenciementData.questions;
      default:
        return [];
    }
  }, [actionEvent]);

  const allSituations = useMemo(() => {
    switch (actionEvent) {
      case MatomoActionEvent.HEURE_RECHERCHE_EMPLOI:
        return heuresRechercheEmploiData.situations;
      case MatomoActionEvent.PREAVIS_DEMISSION:
        return preavisDemissionData.situations;
      case MatomoActionEvent.PREAVIS_LICENCIEMENT:
        return preavisLicenciementData.situations;
      default:
        return [];
    }
  }, [actionEvent]);

  const questionsMap = useMemo(() => {
    return questions.reduce((state, v) => ({ ...state, [v.name]: v }), {});
  }, [questions]);

  const criteriaOrder = useMemo(() => {
    return questions.map(({ name }) => name);
  }, [questions]);

  const memoizedQuestions: any = useMemo(() => {
    const initialSituations = getSituationsFor(allSituations, {
      idcc,
      typeRupture,
    });
    const pastQuestions = getPastQuestions(
      initialSituations,
      criteriaOrder,
      criteria
    );
    const possibleSituations = filterSituations(initialSituations, criteria);
    const nextQuestionKey = getNextQuestionKey(
      possibleSituations,
      criteriaOrder,
      criteria
    );
    const nextQuestionOptions = getOptions(possibleSituations, nextQuestionKey);
    return [...pastQuestions, [nextQuestionKey, nextQuestionOptions]];
  }, [criteria, idcc, allSituations, criteriaOrder]);

  const [isOpenArray, setIsOpenArray] = React.useState<OpenArray>([]);

  const handleChange = (key: string) => {
    let isFound = false;
    const arr = isOpenArray.map((v) => {
      if (v.key === key) {
        isFound = true;
        return { ...v, status: !v.status };
      }
      return v;
    });
    if (!isFound) {
      setIsOpenArray([...isOpenArray, { key, status: true }]);
    } else {
      setIsOpenArray(arr);
    }
  };

  return (
    <>
      <SectionTitle>Statut du salarié</SectionTitle>
      {memoizedQuestions
        .filter(([key]) => !!key)
        .map(([key, answers], index) => (
          <div key={`${index}-${key}-${criteria[key] ? "set" : "unset"}`}>
            {questionsMap[key] && (
              <SelectQuestion
                name={`criteria.${key}`}
                options={answers}
                label={questionsMap[key].question}
                subLabel={subLabel(key)}
                onChange={() => {
                  trackQuestion(questionsMap[key].name, actionEvent, false);
                  form.batch(() => {
                    getFormProps({
                      criteria,
                      key,
                      pastQuestions: memoizedQuestions,
                    }).forEach((key) => {
                      if (key) {
                        form.change(`criteria.${key}`, undefined);
                      }
                    });
                  });
                }}
                isTooltipOpen={
                  isOpenArray.find((v) => v.key === key)?.status ?? false
                }
                onSwitchTooltip={() => handleChange(key)}
                tooltip={
                  questionsMap[key].note !== undefined
                    ? {
                        content: <Html>{questionsMap[key].note}</Html>,
                        trackableFn: (visibility) => {
                          if (visibility) {
                            trackQuestion(questionsMap[key].name, actionEvent);
                          }
                        },
                      }
                    : undefined
                }
              />
            )}
          </div>
        ))}
    </>
  );
}
