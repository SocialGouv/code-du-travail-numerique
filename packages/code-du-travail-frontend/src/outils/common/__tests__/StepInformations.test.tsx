import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { Form } from "react-final-form";

import { generateAgreement } from "../../../../test/generateAgreement";
import { EmbeddedInjectedForm } from "../../../../test/TestForm";
import { MatomoActionEvent } from "../../../lib";
import { StepInformations, StepInformationsProps } from "../StepInformations";
import { FormContent } from "../type/WizardType";

describe("<StepInformations />", () => {
  it("should render", () => {
    const { container } = render(
      <Form initialValues={{}} onSubmit={jest.fn()}>
        {({ form }) => (
          <StepInformations
            actionEvent={MatomoActionEvent.PREAVIS_DEMISSION}
            form={form}
          />
        )}
      </Form>
    );
    expect(container).toMatchSnapshot();
  });

  it("should render questions for specific agreement", () => {
    const { getByText, getAllByRole, getByRole } = render(
      <EmbeddedInjectedForm<FormContent, Omit<StepInformationsProps, "form">>
        Step={StepInformations}
        formData={{
          ccn: { route: "agreement", selected: generateAgreement(1483) },
        }}
        props={{
          actionEvent: MatomoActionEvent.PREAVIS_DEMISSION,
        }}
      />
    );

    expect(getByText(/Statut du salarié/)).toBeInTheDocument();
    expect(
      getByText(/Quelle est la catégorie professionnelle du salarié \?/)
    ).toBeInTheDocument();
    expect(getAllByRole("option").length).toBe(4);
    expect(
      (getByRole("option", { name: "..." }) as HTMLOptionElement).selected
    ).toBeTruthy();
    expect(getByRole("option", { name: "Employés" })).toBeInTheDocument();
    expect(
      getByRole("option", { name: "Agents de maîtrise" })
    ).toBeInTheDocument();
    expect(getByRole("option", { name: "Cadres" })).toBeInTheDocument();
    userEvent.selectOptions(
      // Find the select element
      getByRole("combobox"),
      // Find and select the Ireland option
      getByRole("option", { name: "Employés" })
    );
    expect(
      getByText(/Quelle est l'ancienneté du salarié \?/)
    ).toBeInTheDocument();
    expect(getAllByRole("option").length).toBe(7);
    expect(
      (getAllByRole("option", { name: "..." })[1] as HTMLOptionElement).selected
    ).toBeTruthy();
    expect(
      getByRole("option", { name: "Moins de 6 mois" })
    ).toBeInTheDocument();
    expect(getByRole("option", { name: "Plus de 6 mois" })).toBeInTheDocument();
  });
});
