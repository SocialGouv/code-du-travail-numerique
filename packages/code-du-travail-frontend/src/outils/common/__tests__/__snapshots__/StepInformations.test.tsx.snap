// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StepInformations /> should render 1`] = `
<div>
  <h2
    class="sc-ZqGJI dXpxHJ"
  >
    Statut du salari√©
  </h2>
  <div />
</div>
`;
